src/replication.c

```diff
@@ -1914,6 +1943,169 @@ disconnect_nonresponding_replica(replica_t *replica, uint64_t io_seq,
 	MTX_UNLOCK(&replica->r_mtx);
 }
 
+static void
+handle_snap_list_response(replica_t *replica, void *resp_data, mgmt_cmd_t *mgmt_cmd, uint64_t len)
+{
+	zvol_io_hdr_t *hdr = replica->mgmt_io_resp_hdr;
+	struct zvol_snapshot_list *snap_details;
+	rcommon_mgmt_cmd_t *rcomm_mgmt = mgmt_cmd->rcomm_mgmt;
+	bool delete = false, assigned, failed = false;
+	struct zvol_snapshot_list **snap_resp_array = (struct zvol_snapshot_list **)rcomm_mgmt->buf;
+	uint64_t i = 0;
+
+	snap_details = calloc(1, len);
+	memcpy(snap_details, resp_data, len);
+	free(resp_data);
+
+	MTX_LOCK(&rcomm_mgmt->mtx);
+	if (hdr->status != ZVOL_OP_STATUS_OK || snap_details->error) {
+		rcomm_mgmt->cmds_failed++;
+		free(snap_details);
+		failed = true;
+	} else {
+		rcomm_mgmt->cmds_succeeded++;
+	}
+
+	if ((rcomm_mgmt->caller_gone == 1) &&
+	    (rcomm_mgmt->cmds_sent == (rcomm_mgmt->cmds_failed + rcomm_mgmt->cmds_succeeded)))
+		delete = true;
+
+	if (failed)
+		goto out;
+
+	for (i = 0, assigned = false;
+	    i < (rcomm_mgmt->buf_size / sizeof (*snap_resp_array));
+		i++) {
+		if (snap_resp_array[i] == NULL) {
+			snap_resp_array[i] = (struct zvol_snapshot_list *)snap_details;
+			assigned = true;
+			break;
+		}
+	}
+
+	ASSERT(assigned);
+
+out:
+
+	MTX_UNLOCK(&rcomm_mgmt->mtx);
+
+	if (delete == true) {
+		DESTROY_SNAPSHOT_RESP_LIST(rcomm_mgmt);
+		free_rcommon_mgmt_cmd(rcomm_mgmt);
+	}
+}
+
+static char *
+get_replica_id_for_guid(spec_t *spec, uint64_t guid)
+{
+	replica_t *replica;
+	char *replica_id = NULL;
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica->zvol_guid == guid) {
+			replica_id = xstrdup(replica->replica_id);
+			break;
+		}
+	}
+
+	return replica_id;
+}
+
+static void
+process_snaplist_response(spec_t *spec, char **resp, void *buf, int rcount)
+{
+	struct zvol_snapshot_list **snap_resp_array = (struct zvol_snapshot_list **)buf;
+	int i = 0;
+	uint64_t total_len;
+	char *msg = NULL, *replica_id = NULL;
+	const char *json_string = NULL;
+	struct json_object *jobj, *jarray, *resp_obj, *snap_obj;
+
+	jarray = json_object_new_array();
+
+	for (i = 0; i < rcount; i += 1) {
+		if (snap_resp_array[i]) {
+			resp_obj = snap_obj = NULL;
+			jobj = json_object_new_object();
+
+			replica_id = get_replica_id_for_guid(spec, snap_resp_array[i]->zvol_guid);
+			if(!replica_id) {
+				REPLICA_ERRLOG("No replica_id for replica:%lu", snap_resp_array[i]->zvol_guid);
+				continue;
+			}
+
+			json_object_object_add(jobj, "replica_id", json_object_new_string(replica_id));
+			if (snap_resp_array[i]->error)
```

Comment:
is this case possible? we are not putting in the response struct if there is an error. probably an assert for that would be good?

Reply:
done